app/models/grda_warehouse/tasks/identify_duplicates_query_matcher.rb

```diff
@@ -0,0 +1,241 @@
+      # @return [Array<Array<Integer>>] Array of ID pairs, or empty array if no unprocessed_ids
+      def execute(warehouse_id: nil, validate_ssn: nil)
+        # Default validate_ssn to true for SSN matches
+        validate_ssn = (field_alias == 'ssn') if validate_ssn.nil?
+
+        # Early return for unprocessed with no IDs
+        return [] if match_type == :unprocessed && unprocessed_ids.blank?
+
+        results = GrdaWarehouse::Hud::Client.connection.execute(to_sql(warehouse_id: warehouse_id))
+
+        # Filter SSNs if validation requested and SSN is in results
+        results = results.select { |r| ::HudHelper.util.valid_social?(r['ssn']) } if validate_ssn && include_value_in_results && field_alias == 'ssn'
+
+        # Map to ID pairs based on match type
+        case match_type
+        when :existing
+          results.map { |r| [r['destination_one_id'], r['destination_two_id']] }.uniq
+        when :unprocessed
+          results.map { |r| [r['destination_client_id'], r['source_client_id']] }.uniq
+        end
+      end
+
+      private
+
+      attr_reader :match_type, :field_expression, :field_alias, :field_filters, :include_value_in_results,
+                  :destination_data_source_ids, :unprocessed_ids, :warehouse_id
+
+      # Space-efficient approach: Groups all matching clients by field value into arrays,
+      # then generates unique pairs using array indices. This avoids materializing
+      # the full cartesian product (N² temporary rows) produced by self-joins.
+      def existing_sql
+        <<-SQL
+          WITH grouped_matches AS (
+            SELECT
+              #{field_expression} AS #{field_alias},
+              array_agg(DISTINCT warehouse_clients.destination_id ORDER BY warehouse_clients.destination_id) AS destination_ids
+            #{existing_from_clause}
+            GROUP BY #{field_alias}
+            HAVING COUNT(DISTINCT warehouse_clients.destination_id) > 1
+          )
+          SELECT
+            destination_ids[idx_one] AS destination_one_id,
+            destination_ids[idx_two] AS destination_two_id#{value_projection}
+          FROM grouped_matches
+          CROSS JOIN LATERAL (
+            SELECT idx_one, idx_two
+            FROM generate_subscripts(destination_ids, 1) AS idx_one
+            JOIN generate_subscripts(destination_ids, 1) AS idx_two ON idx_one < idx_two
+          ) AS destination_pairs
+        SQL
+      end
+
+      def existing_from_clause
+        filters = base_filters + field_filters
+        <<-SQL
+          FROM "Client" AS clients
+          INNER JOIN warehouse_clients ON clients.id = warehouse_clients.source_id
+          WHERE #{filters.join("\n            AND ")}
+        SQL
+      end
+
+      def unprocessed_sql
+        destination_ids_sql = destination_data_source_ids&.join(', ')
+        unprocessed_ids_sql = unprocessed_ids&.join(', ')
```

Comment:
Are we concerned that these might contain hundreds of thousands of ids or more?  I think that's why we went to temp tables, but I might be mis-remembering.